### PR TITLE
Added failOnStdError input

### PR DESCRIPTION
**Pull request by t-dedah**
Added this input to support dependency on actions which are passing warning in stderr
Discussion based on this issue https://github.com/Azure/arm-deploy/issues/59

### CODE REVIEW

**Comment by alex-frankel** [comment]
@t-dedah - if I use this input, will we fall back to relying on the error code? Otherwise, won't it be impossible for this task to fail?

File: README.md [hunk 1]
```diff
@@ -20,8 +20,7 @@ This action can be used to deploy Azure Resource Manager templates at different
 * `parameters`: Specify the path or URL to the Azure Resource Manager deployment parameter values. Or local / remote value file.  
 * `deploymentMode`: `Incremental`(default) (only add resources to resource group) or `Complete` (remove extra resources from resource group) or `Validate` (only validates the template). 
 * `deploymentName`: Specifies the name of the resource group deployment to create.
-
-
+* `failOnStdError`: Specify whether to fail the action when std error is not null.
```
Comment (bishal-pdMSFT) [inline README.md:23]: Rename to failOnStdErr as stderr is more easily recognised term

Comment (bishal-pdMSFT) [reply README.md:23]: Use that in description as well

Comment (bishal-pdMSFT) [inline README.md:23]: Specify the default value

File: README.md [hunk 2]
```diff
@@ -20,8 +20,7 @@ This action can be used to deploy Azure Resource Manager templates at different
 * `parameters`: Specify the path or URL to the Azure Resource Manager deployment parameter values. Or local / remote value file.  
 * `deploymentMode`: `Incremental`(default) (only add resources to resource group) or `Complete` (remove extra resources from resource group) or `Validate` (only validates the template). 
 * `deploymentName`: Specifies the name of the resource group deployment to create.
-
-
```
Comment (bishal-pdMSFT) [inline README.md:24]: Add an example for stderr and explain what it means via the example

File: README.md (hunk 1 above)
Comment (bishal-pdMSFT) [inline README.md:23]: Rather than saying not null, say if the some data is written to stderr stream of az cli

File: src/deploy/scope_managementgroup.ts [hunk 3]
```diff
@@ -68,7 +67,11 @@ export async function DeployManagementGroupScope(azPath: string, region: string,
         core.info("Creating deployment...")
         var deploymentCode = await exec(`"${azPath}" deployment mg create ${azDeployParameters} -o json`, [], deployOptions);
         if (commandStdErr.trim().length !== 0) {
-            throw new Error(`Deployment process failed as some lines were written to stderr: ${commandStdErr}`)
```
Comment (bishal-pdMSFT) [inline src/deploy/scope_managementgroup.ts:71]: Repeated block of code. Please see if it can be moved to utility or common module

File: src/deploy/scope_managementgroup.ts [hunk 4]
```diff
@@ -68,7 +67,11 @@ export async function DeployManagementGroupScope(azPath: string, region: string,
...
+            if (failOnStdError.toLowerCase().trim() == "true") {
```
Comment (bishal-pdMSFT) [inline src/deploy/scope_managementgroup.ts:70]: Pass failOnStdError as a boolean here. To the string trim and comparison in entry point only

File: src/deploy/scope_managementgroup.ts [hunk 5]
```diff
@@ -68,7 +67,11 @@ export async function DeployManagementGroupScope(azPath: string, region: string,
...
+                throw new Error(`Deployment process failed as some lines were written to stderr: ${commandStdErr}`)
+            } else {
+                core.error(commandStdErr)
+            }
         } else {
```
Comment (bishal-pdMSFT) [inline src/deploy/scope_managementgroup.ts:75]: Else should not be there anymore otherwise it will not check for exit code and will not parse outputs

Comment (bishal-pdMSFT) [inline src/deploy/scope_managementgroup.ts:75]: Please add test for the negative case

File: src/main.ts [hunk 6]
```diff
@@ -22,6 +22,7 @@ export async function main(): Promise<Outputs> {
     const deploymentName = getInput('deploymentName')
     const parameters = getInput('parameters')
     const managementGroupId = getInput('managementGroupId')
+    const failOnStdError = getInput('failOnStdError')
```
Comment (bishal-pdMSFT) [inline src/main.ts:25]: Convert to boolean here itself

**Comment by t-dedah** [comment]
> @t-dedah - if I use this input, will we fall back to relying on the error code? Otherwise, won't it be impossible for this task to fail?

@alex-frankel Now we are first checking exit code and then failOnStdErr input.

1.  If exit code is 1 then we directly fail the action
2.  if exit code is 0 and stdErr is empty then action is passed
3. if exit code is 0, stdErr is non-empty and failOnStdErr is set to true(default value ) then action will fail
4.  if exit code is 0, stdErr is non-empty and failOnStdErr is set to false then action will pass

**Comment by alex-frankel** [comment]
So if the exit code is 1, the action will fail -- regardless of the value of `failOnStdErr`? That looks like the right behavior to me.

**Comment by t-dedah** [comment]
Yes @alex-frankel

File: src/deploy/scope_managementgroup.ts [hunk 7]
```diff
@@ -67,18 +66,23 @@ export async function DeployManagementGroupScope(azPath: string, region: string,
         // execute the deployment
         core.info("Creating deployment...")
         var deploymentCode = await exec(`"${azPath}" deployment mg create ${azDeployParameters} -o json`, [], deployOptions);
-        if (commandStdErr.trim().length !== 0) {
-            throw new Error(`Deployment process failed as some lines were written to stderr: ${commandStdErr}`)
-        } else {
-            if (deploymentCode != 0) {
-                core.error("Deployment failed.")
+        
+        if (deploymentCode == 0 && commandStdErr.trim().length !== 0) {
```
Comment (bishal-pdMSFT) [inline src/deploy/scope_managementgroup.ts:70]: can this be simplified by:

First log the error wherever they get into stderr listener and set a flag as well

```
stderr: (data: BufferSource) => {
                error = data.toString();
                if(error && error.trim().length !== 0)
                {
                    commandStdErr = true;
                    core.error(error);
                }
            }
```


```
        if (deploymentCode == 0) {
            throw new Error("Deployment failed.")
        }
        if(commandStdErr && failOnStdErr) {
            throw new Error("Deployment process failed as some lines were written to stderr");
        }
        
        core.debug(commandOutput);
        core.info("Parsing outputs...")
        return ParseOutputs(commandOutput)
```

File: README.md [hunk 8]
```diff
@@ -140,6 +139,25 @@ If we now add a Shell script with a simple echo from that value, we can see that
...
+- uses: azure/arm-deploy@v1
+  id: deploy
+  with:
+    resourceGroupName: azurearmaction
+    template: examples/template/template.json
+    parameters: examples/template/parameters.json
+    deploymentName: github-advanced-test
+    failOnStdErr: false
+```
+The default value of `failOnStdErr` is true to support back compatibility. `failOnStdErr` equals true means that if there is some data written to stdErr and return code from az-cli is 0 then we will fail the action.
```
Comment (bishal-pdMSFT) [inline README.md:155]: Remove this `The default value of `failOnStdErr` is true to support back compatibility.`

File: src/main.ts [hunk 9]
```diff
@@ -22,24 +22,38 @@ export async function main(): Promise<Outputs> {
     const deploymentName = getInput('deploymentName')
     const parameters = getInput('parameters')
     const managementGroupId = getInput('managementGroupId')
+    const failOnStdErr = getInput('failOnStdErr')
```
Comment (bishal-pdMSFT) [inline src/main.ts:25]: you should be able to use `core.getBooleanInput()` function: https://github.com/actions/toolkit/tree/main/packages/core#inputsoutputs

This will remove need to do conversion later in L33

Comment (bishal-pdMSFT) [reply src/main.ts:25]: What I am not sure is if it will take automatically default value specified in action.yml if user does not specify it?

Comment (bishal-pdMSFT) [reply src/main.ts:25]: If not so, then you can `|| true`

File: .github/workflows/ci-workflow.yml [hunk 10]
```diff
@@ -37,6 +51,7 @@ jobs:
           INPUT_TEMPLATE: ./test/template.json
           INPUT_PARAMETERS: ./test/parameters.json
           INPUT_DEPLOYMENTNAME: github-test-mg
+          EXPECTED_TO: pass
```
Comment (bishal-pdMSFT) [inline .github/workflows/ci-workflow.yml:54]: This should be optional and need not be specified for `pass`.

File: src/main.ts (hunk 9 above)
Comment (t-dedah) [reply src/main.ts:25]: @bishal-pdMSFT This method is not available in the current version of `@actions/core` which we are using. Should we update the package then?

File: README.md [hunk 11]
```diff
@@ -140,6 +139,25 @@ If we now add a Shell script with a simple echo from that value, we can see that
...
+- uses: azure/arm-deploy@v1
+  id: deploy
+  with:
+    resourceGroupName: azurearmaction
+    template: examples/template/template.json
+    parameters: examples/template/parameters.json
+    deploymentName: github-advanced-test
+    failOnStdErr: false
+```
+`failOnStdErr` equals true means that if there is some data written to stdErr and return code from az-cli is 0 then we will fail the action.
```
Comment (bishal-pdMSFT) [inline README.md:155]:
```suggestion
`failOnStdErr` equals true implied that if some data is written to stdErr and exit code from az-cli is 0, then action execution will fail. 
```

File: README.md [hunk 12]
```diff
@@ -140,6 +139,25 @@ If we now add a Shell script with a simple echo from that value, we can see that
...
+  with:
+    resourceGroupName: azurearmaction
+    template: examples/template/template.json
+    parameters: examples/template/parameters.json
+    deploymentName: github-advanced-test
+    failOnStdErr: false
+```
+`failOnStdErr` equals true means that if there is some data written to stdErr and return code from az-cli is 0 then we will fail the action. 
+
+`failOnStdErr` equals false means that if there is some data written to stdErr and return code from az-cli is 0 then action will be return a success. This input is added to support cases where stdErr is being used to stream warnings.
```
Comment (bishal-pdMSFT) [inline README.md:157]:
```suggestion
`failOnStdErr` equals false implies that if some data is written to stdErr and return code from az-cli is 0, then action will continue execution. This input is added to support cases where stdErr is being used to stream warning or progress info. 
```

File: README.md [hunk 13]
```diff
@@ -140,6 +139,25 @@ If we now add a Shell script with a simple echo from that value, we can see that
...
+    template: examples/template/template.json
+    parameters: examples/template/parameters.json
+    deploymentName: github-advanced-test
+    failOnStdErr: false
+```
+`failOnStdErr` equals true means that if there is some data written to stdErr and return code from az-cli is 0 then we will fail the action. 
+
+`failOnStdErr` equals false means that if there is some data written to stdErr and return code from az-cli is 0 then action will be return a success. This input is added to support cases where stdErr is being used to stream warnings. 
+
+Return code 1 will always lead to failure of action irrespective the value of `failOnStdErr`.
```
Comment (bishal-pdMSFT) [inline README.md:159]:
```suggestion
Non zero Exit code will always lead to failure of action irrespective the value of `failOnStdErr`.
```

**Review by bishal-pdMSFT — APPROVED** [review]
